testsuites/CBLTester/CBL_Functional_tests/TestSetup_FunctionalTests/test_replication.py

```diff
@@ -4090,6 +4208,272 @@ def verify_sgDocIds_cblDocIds(sg_client, url, sg_db, session, cbl_db, db):
+    sg_url = params_from_base_test_setup["sg_url"]
+    sg_admin_url = params_from_base_test_setup["sg_admin_url"]
+    cluster_config = params_from_base_test_setup["cluster_config"]
+    sg_blip_url = params_from_base_test_setup["target_url"]
+    base_url = params_from_base_test_setup["base_url"]
+    sg_config = params_from_base_test_setup["sg_config"]
+    db = params_from_base_test_setup["db"]
+    cbl_db = params_from_base_test_setup["source_db"]
+    sync_gateway_version = params_from_base_test_setup["sync_gateway_version"]
+
+    if sync_gateway_version < "2.0.0":
+        pytest.skip('This test cannot run with sg version below 2.0')
+    channels_sg = ["replicator"]
+    username = "autotest"
+    password = "password"
+
+    # Create CBL database
+    sg_client = MobileRestClient()
+
+    # Reset cluster to ensure no data in system
+    c = cluster.Cluster(config=cluster_config)
+    c.reset(sg_config_path=sg_config)
+    sg_controller = SyncGateway()
+
+    # Configure replication with push_pull
+    replicator = Replication(base_url)
+    sg_client.create_user(sg_admin_url, sg_db, username, password, channels=channels_sg)
+    authenticator = Authenticator(base_url)
+    cookie, session_id = sg_client.create_session(sg_admin_url, sg_db, username)
+    replicator_authenticator = authenticator.authentication(session_id, cookie, authentication_type="session")
+
+    repl_config = replicator.configure(cbl_db, sg_blip_url, continuous=continuous, channels=channels_sg,
+                                       max_retry_wait_time=wait_time, max_retries=retries,
+                                       replication_type=type, replicator_authenticator=replicator_authenticator)
+    repl = replicator.create(repl_config)
+    db.create_bulk_docs(num_of_docs, "cbl-replicator-retries", db=cbl_db, channels=channels_sg)
+
+    start_time = time.time()
+    repl_change_listener = replicator.addChangeListener(repl)
+
+    # Stop Sync Gateway
+    sg_controller = SyncGateway()
+
+    sg_controller.stop_sync_gateways(cluster_config, url=sg_url)
+    end_time = time.time()
+    replicator.start(repl)
+    sg_controller.start_sync_gateways(cluster_config, url=sg_url, config=sg_config)
+    changes_count = replicator.getChangesCount(repl_change_listener)
+    log_info("*" * 90)
+    time_taken = end_time - start_time
+    log_info(time_taken)
+    log_info(changes_count)
+    assert changes_count > 20
+    replicator.wait_until_replicator_idle(repl)
+
+    # Stop the sg and restart the replicator
+    sg_controller.stop_sync_gateways(cluster_config, url=sg_url)
+
+    # start the sg before retries ends
+    sg_controller.start_sync_gateways(cluster_config, url=sg_url, config=sg_config)
+    # Added to complete number of iteration
+    time.sleep(8)
+    replicator.wait_until_replicator_idle(repl)
+
+    sg_docs = sg_client.get_all_docs(url=sg_admin_url, db=sg_db, include_docs=True)
+    sg_docs = sg_docs["rows"]
+    # Verify database doc counts
```

Comment:
Suggestion : you can add steps as comments, which will be easy to go through the test which matches with the steps